packages/store/addon/-private/system/fetch-manager.ts

```diff
@@ -0,0 +1,554 @@
+    // used to keep track of all the find requests that need to be coalesced
+    this._pendingFetch = new Map();
+    this._adapterCache = adapterCache;
+    this._store = store;
+
+    this._pendingSave = [];
+    this.requestCache = new RequestCache();
+  }
+
+  /**
+    This method is called by `record.save`, and gets passed a
+    resolver for the promise that `record.save` returns.
+
+    It schedules saving to happen at the end of the run loop.
+
+    @method scheduleSave
+    @private
+    @param {InternalModel} internalModel
+    @param {Resolver} resolver
+    @param {Object} options
+  */
+  scheduleSave(identifier: RecordIdentifier, options: any = {}) {
+    let promiseLabel = 'DS: Model#save ' + this;
+    let resolver = RSVP.defer(promiseLabel);
+    let query: SaveRecordMutation = {
+      'op': 'saveRecord',
+      recordIdentifier: identifier,
+      options
+    }
+
+    let queryRequest: Request = {
+      data: [query]
+    }
+
+    let snapshot = new Snapshot(options, identifier, this._store);
+    let pendingSaveItem = {
+      snapshot: snapshot,
+      resolver: resolver,
+      identifier,
+      options,
+      queryRequest
+    }
+    this._pendingSave.push(pendingSaveItem);
+    emberRun.scheduleOnce('actions', this, this._flushPendingSaves);
+
+    this.requestCache.enqueue(resolver.promise, pendingSaveItem.queryRequest);
+
+    return resolver.promise;
+  }
+
+  _flushPendingSave(pending: PendingSaveItem) {
+    let { snapshot, resolver, identifier } = pending;
+    let adapter = this._store.adapterFor(identifier.type);
+    let operation;
+    let recordData: RelationshipRecordData = this._store.recordDataForIdentifier(identifier);
+
+    if (recordData.isNew()) {
+      operation = 'createRecord';
+    } else if (recordData.isDeleted()) {
+      operation = 'deleteRecord';
+    } else {
+      operation = 'updateRecord';
+    }
+    let internalModel = snapshot._internalModel;
+    let modelName = snapshot.modelName;
+    let store = this._store;
+    let modelClass = store.modelFor(modelName);
+
+    assert(`You tried to update a record but you have no adapter (for ${modelName})`, adapter);
+    assert(
+      `You tried to update a record but your adapter (for ${modelName}) does not implement '${operation}'`,
+      typeof adapter[operation] === 'function'
+    );
+
+    let promise = Promise.resolve().then(() => adapter[operation](store, modelClass, snapshot));
+    let serializer = serializerForAdapter(store, adapter, modelName);
+    let label = `DS: Extract and notify about ${operation} completion of ${internalModel}`;
+
+    assert(
+      `Your adapter's '${operation}' method must return a value, but it returned 'undefined'`,
+      promise !== undefined
+    );
+
+    promise = guardDestroyedStore(promise, store, label);
+    promise = _guard(promise, _bind(_objectIsAlive, internalModel));
+
+    promise = promise.then(
+      adapterPayload => {
+        let payload, data, sideloaded;
```

Comment:
needs type annotations x3

Reply:
legacy code that was c/ped so will type as a followup